[Serve][2/N] Add deployment-level autoscaling snapshot and event summarizer

nadongjun opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR introduces deployment-level autoscaling observability in Serve. The controller now emits a single, structured JSON log line (serve_autoscaling_snapshot) per autoscaling-enabled deployment each control-loop tick.

This avoids recomputation in the controller call sites and provides a stable, machine-parsable surface for tooling and debugging.

#### Changed
- Add get_observability_snapshot in AutoscalingState and manager wrapper to generate compact snapshots (replica counts, queued/total requests, metric freshness).
- Add ServeEventSummarizer to build payloads, reduce duplicate logs, and summarize recent scaling decisions.

#### Example log (single line):
Logs can be found in controller log files, `e.g. /tmp/ray/session_2025-09-03_21-12-01_095657_13385/logs/serve/controller_13474.log`.

```
serve_autoscaling_snapshot {"ts":"2025-09-04T06:12:11Z","app":"default","deployment":"worker","current_replicas":2,"target_replicas":2,"replicas_allowed":{"min":1,"max":8},"scaling_status":"stable","policy":"default","metrics":{"look_back_period_s":10.0,"queued_requests":0.0,"total_requests":0.0},"metrics_health":"ok","errors":[],"decisions":[{"ts":"2025-09-04T06:12:11Z","from":0,"to":2,"reason":"current=0, proposed=2"},{"ts":"2025-09-04T06:12:11Z","from":2,"to":2,"reason":"current=2, proposed=2"}]}
```
#### Follow-ups
- Expose the same snapshot data via `serve status -v` and CLI/SDK surfaces.
- Aggregate per-app snapshots and external scaler history.

## Related issue number

#55834

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

nadongjun committed:
[Serve] Add deployment-level autoscaling snapshot and event summarizer

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Add autoscaling snapshot log test

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
rafactor and lint

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
+                    )
+                    self._serve_event_summarizer.log_snapshot(payload)
+                    self._last_autoscaling_snapshots[key] = signature
+        except Exception:
+            # Never let logging issues impact the control loop.
+            logger.exception("Failed emitting autoscaling snapshot.")
```
handling exception here is unnecessary, since caller already does it.

abrarsheikh commented on python/ray/serve/_private/logging_utils.py:
```diff
@@ -517,3 +520,155 @@ def __enter__(self):
             self.logger.setLevel(self.old_level)
+
+
+class ServeEventSummarizer:
```
```suggestion
class ServeAutoscalingEventSummarizer:
```

abrarsheikh replied:
i think this class should be moved to its own file and not be part of `logging_utils`, its responsibility doesn't seem too related to what `logging_utils`  does

abrarsheikh commented on python/ray/serve/_private/logging_utils.py:
```diff
@@ -517,3 +520,155 @@ def __enter__(self):
+            "serve_autoscaling_snapshot " + json.dumps(payload, separators=(",", ":"))
+        )
+
+    def note_once_per_interval(
```
who calls this?

abrarsheikh commented on python/ray/serve/_private/logging_utils.py:
```diff
@@ -517,3 +520,155 @@ def __enter__(self):
+
+    def __init__(self, logger: logging.Logger):
+        self._summarizer = EventSummarizer()
+        self._logger = logger
```
use global logger instead of passing logger into the class. This is not a pattern we have used in the repo.

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
+                    )
+
+                    self._scaling_decisions.setdefault(dep_id, []).append(decision)
+                    if len(self._scaling_decisions[dep_id]) > 50:
```
move 50 to a constant.

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
         new_proxy_nodes.add(self._controller_node_id)
         self._proxy_nodes = new_proxy_nodes
 
+    def _emit_deployment_autoscaling_snapshots(self) -> None:
```
this function is too complex, let's break it into smaller units

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
+                    autoscaling_config = details.deployment_config.autoscaling_config
+                    if not autoscaling_config:
+                        continue
+                    current_replicas = len(getattr(details, "replicas", []) or [])
```
`details.replicas` will always be a list. Wondering why use defensive getattr?

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
+                    if not autoscaling_config:
+                        continue
+                    current_replicas = len(getattr(details, "replicas", []) or [])
+                    raw_target = getattr(details, "target_num_replicas", None)
```
same comment here `details.target_num_replicas` cannot be None based on pydantic model

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
+                        continue
+                    current_replicas = len(getattr(details, "replicas", []) or [])
+                    raw_target = getattr(details, "target_num_replicas", None)
+                    current_target = (
+                        int(raw_target) if raw_target is not None else current_replicas
+                    )
```
`raw_target` cannot be None, i think.

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
+                    max_repl_adj = snapshot["max_replicas"]
+                    total_requests = snapshot["total_requests"]
+
+                    if isinstance(autoscaling_config, dict):
+                        policy_name = autoscaling_config.get("policy", "default")
+                        look_back_period_s = autoscaling_config.get(
+                            "look_back_period_s"
+                        )
+                    else:
+                        policy_name = getattr(autoscaling_config, "name", "default")
+                        look_back_period_s = getattr(
+                            autoscaling_config, "look_back_period_s", None
+                        )
```
this seems odd, why can't we assume `autoscaling_config` as either a dict or pydantic object? why do we need this dual logic for accessing attributes of autoscaling_config

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
+                            scaling_status
+                        )
+                    )
+                    payload = self._serve_event_summarizer.build_snapshot_payload(
```
do you expect payload to be used for anything other than `log_snapshot`? Why not log it as part of same function?

abrarsheikh reviewed (commented):
my main feedback about this PR is that we are creating many intermediate free form dictionaries, and it is not clear to me why we need them all but importantly they create ambiguity in future about what each dictionary is supposed to contain making maintaining code harder. The code can be reorganized better, used typed objects to function that need to return large dictionaries.

nadongjun committed:
[comment 1, 2, 4] Replace dicts with typed objects, remove redundant except and unused func(note_once_per_interval)

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
[comment 3, 5, 6] Move/rename autoscaling summarizer, use global logger, and add constant

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
[comment 8, 9] Add InternalAutoscalingConfig for normalized type and remove unnecessary getattr

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
[comment 10] use to_log_dict() and emit_deployment_snapshot()

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Add constant(AUTOSCALER_SUMMARIZER_DECISION_LIMIT)

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
[comment 7] Refactor _emit_deployment_autoscaling_snapshots

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
lint

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
lint

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

akyang-anyscale commented on python/ray/serve/_private/event_summarizer.py:
```diff
@@ -0,0 +1,198 @@
+    """
+
+    def __init__(self):
+        self._summarizer = EventSummarizer()
```
let's remove this if it's not used

akyang-anyscale commented on python/ray/serve/_private/event_summarizer.py:
```diff
@@ -0,0 +1,198 @@
+        look_back_period_s: Optional[float],
+    ) -> str:
+        """Return a short human-friendly health string based on freshness."""
+        if last_metrics_age_s is None:
+            return "ok"
```
does this mean the metrics is missing?

nadongjun replied:
I’ve updated it to return "unknown" in that case, to indicate that the metrics are missing.

akyang-anyscale commented on python/ray/serve/_private/event_summarizer.py:
```diff
@@ -0,0 +1,198 @@
+        """Return a short human-friendly health string based on freshness."""
+        if last_metrics_age_s is None:
+            return "ok"
+        threshold = look_back_period_s or 30.0
+        if last_metrics_age_s > threshold:
+            return f"delayed (last update {int(last_metrics_age_s)}s ago)"
+        return "ok"
```
thoughts on just returning the age instead of `ok` vs `delayed` which seems somewhat arbitrary?

akyang-anyscale commented on python/ray/serve/_private/event_summarizer.py:
```diff
@@ -0,0 +1,198 @@
+        app_name: str,
+        deployment_name: str,
+        current_replicas: int,
+        proposed_replicas: int,
```
```suggestion
        target_replicas: int,
```

akyang-anyscale commented on python/ray/serve/_private/event_summarizer.py:
```diff
@@ -0,0 +1,198 @@
+            timestamp_s=timestamp_s,
+            app=app_name,
+            deployment=deployment_name,
+            current_replicas=int(current_replicas),
+            target_replicas=int(proposed_replicas),
+            min_replicas=None if min_replicas is None else int(min_replicas),
+            max_replicas=None if max_replicas is None else int(max_replicas),
+            scaling_status=scaling_status,
+            policy=policy_name,
+            look_back_period_s=look_back_period_s,
+            queued_requests=None if queued_requests is None else float(queued_requests),
+            total_requests=float(total_requests or 0.0),
```
we should be able to just reference the variables without the casting and None checks

akyang-anyscale commented on python/ray/serve/_private/event_summarizer.py:
```diff
@@ -0,0 +1,198 @@
+        queued_requests: Optional[float],
+        total_requests: float,
+        last_metrics_age_s: Optional[float],
+        errors: Optional[List[str]],
```
```suggestion
        errors: List[str],
```

akyang-anyscale commented on python/ray/serve/_private/event_summarizer.py:
```diff
@@ -0,0 +1,198 @@
+        recent_decisions: List[DecisionSummary],
+    ) -> None:
+        """Build and immediately log a deployment snapshot."""
+        snapshot = self.build_deployment_snapshot(
```
Are there plans to use `build_deployment_snapshot` elsewhere in the future? If not can we just move that method's implementation into here?

nadongjun replied:
I removed build_deployment_snapshot since there were no other callers and no clear plan to reuse it.

akyang-anyscale commented on python/ray/serve/_private/event_summarizer.py:
```diff
@@ -0,0 +1,198 @@
+            "serve_autoscaling_snapshot " + json.dumps(payload, separators=(",", ":"))
+        )
+
+    def emit_deployment_snapshot(
```
```suggestion
    def log_deployment_snapshot(
```

akyang-anyscale commented on python/ray/serve/_private/controller.py:
```diff
@@ -223,6 +230,13 @@ async def __init__(
         self._proxy_nodes = set()
         self._update_proxy_nodes()
 
+        # Serve-side summarizer to throttle repetitive notes and build payloads.
```
```suggestion
        # Serve-side summarizer to throttle repetitive logs and build payloads.
```

akyang-anyscale commented on python/ray/serve/_private/autoscaling_state.py:
```diff
@@ -366,6 +366,65 @@ def get_total_num_requests(self) -> float:
 
         return total_requests
 
+    def get_observability_snapshot(
```
```suggestion
    def get_snapshot(
```

akyang-anyscale commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +389,134 @@ def _update_proxy_nodes(self):
+            decisions_summary = self._serve_event_summarizer.summarize_recent_decisions(
+                recent_decisions
+            )
+            scaling_status_fmt = self._serve_event_summarizer.format_scaling_status(
+                scaling_status
+            )
```
i would push this down to `emit_deployment_snapshot`

akyang-anyscale commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +389,134 @@ def _update_proxy_nodes(self):
+            self._record_decision(
+                dep_id, current_replicas, proposed_replicas, policy_name
+            )
+            key = (app_name, dep_name)
```
can use deployment_id as the key

akyang-anyscale commented on python/ray/serve/_private/controller.py:
```diff
@@ -375,6 +389,134 @@ def _update_proxy_nodes(self):
+                look_back_period_s,
+                snapshot,
+                scaling_status,
+            ) = self._compute_snapshot_inputs(
```
can we directly call `get_observability_snapshot` and use the snapshot object directly

akyang-anyscale reviewed (commented):
Thanks for the contribution @nadongjun! Have you thought about how this would change/work with application-level autoscaling which is in flight: https://github.com/ray-project/ray/pull/56149? When application-level autoscaling is enabled, deployment does not autoscale by-itself, so that may change how user should interpret the logs.

As a feedback for the PR, I would recommend packaging the various autoscaling relevant values into objects, and pass that object around. It's somewhat difficult to track all the different variables and where they come from, and makes the code a bit harder to parse.

nadongjun committed:
[review] Refactor autoscaling snapshot API and schema

- Rename get_observability_snapshot → get_snapshot for clarity
- Rename proposed_replicas → target_replicas across snapshot flow
- Return last_metrics_age_s=None when no metrics; map to "unknown" in summarizer
- Flatten replicas_allowed{min,max} into top-level min, max in snapshot payload
- Move look_back_period_s to top-level for consistency
- Rename DecisionSummary → AutoscalingDecisionSummary for clarity
- Replace tuple-based SnapshotSignature with typed dataclass
- Use DeploymentID directly as dedupe key instead of (app_name, dep_name)
- Inline snapshot computation in controller; remove _compute_snapshot_inputs
- Push scaling_status formatting into log_deployment_snapshot
- Update tests to validate new payload shape (min/max, no replicas_allowed)

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented on python/ray/serve/_private/common.py:
```diff
@@ -754,3 +754,89 @@ class CreatePlacementGroupRequest:
+                "queued_requests": self.queued_requests,
+                "total_requests": self.total_requests,
+            },
+            "metrics_health": self.metrics_health,
```
It indicates the overall health of metrics for the deployment. If there are specific metric errors, they will be added to the errors field.

nadongjun committed:
[review] Snapshot timestamp, metrics age formatting

- Standardize payload to return 'timestamp_s' for snapshots.

- Return metrics health as last_metrics_age_s

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
[review] Inline build_deployment_snapshot into log_deployment_snapshot

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented:
@abrarsheikh @akyang-anyscale Thanks for the detailed review!

@akyang-anyscale That’s a fair point. serve_autoscaling_snapshot log format currently only covers deployment-level autoscaling. Once application-level autoscaling is added, we’ll log deployment and application-level snapshots separately.

I’ve already switched to typed dataclasses (e.g., DeploymentSnapshot, AutoscalingDecisionSummary) so the controller passes structured objects instead of dicts. I’ll do the same for application-level autoscaling to keep things consistent.

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun committed:
Refactor improve readability

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Refactor improve readability

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Update comments and var name

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Refactor deployment snapshot construction

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Add cache deployment snapshot in AutoscalingState

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

akyang-anyscale reviewed (commented):
@nadongjun Let's rethink where we want to create `DeploymentSnapshot` (which we should also probably rename to `DeploymentAutoscalingSnapshot`). We can probably achieve a cleaner implementation if `AutoscalingState` creates a new `DeploymentSnapshot` every autoscaling decision and caches that snapshot. The information needed to create a new `DeploymentSanpshot` should all be available to `AutoscalingState` (aside from `_scaling_decisions` currently which we can also add to each `AutoscalingState`).

When the controller emits autoscaling logs, it can then pull the current `DeploymentSnapshot` for each deployment from the `AutoscalingStateManager`, and log it if it has changed since the last time we've logged. This way, all logic corresponding to creating a `DeploymentSnapshot` is contained to `AutoscalingState`, and controller only has to handle the logging logic.

nadongjun committed:
Update python/ray/serve/_private/autoscaling_state.py

Co-authored-by: akyang-anyscale <alexyang@anyscale.com>
Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Update python/ray/serve/_private/controller.py

Co-authored-by: akyang-anyscale <alexyang@anyscale.com>
Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Update python/ray/serve/_private/controller.py

Co-authored-by: akyang-anyscale <alexyang@anyscale.com>
Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Remove SnapshotSignature and implement __eq__

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Add ApplicationStateManager.list_app_names()

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Remove normalize_autoscaling_config

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Move snapshot creation to AutoscalingState

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Refactor get_deployment_snapshot to return cached value only

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Refactor autoscaling snapshot handling:
- Make _create_deployment_snapshot return a fully populated DeploymentSnapshot from AutoscalingState
- Move scaling decision history into each AutoscalingState (remove controller-owned tracking)
- Integrate summarization/formatting into DeploymentSnapshot and remove ServeAutoscalingEventSummarizer

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
lint

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun committed:
Merge remote-tracking branch 'upstream/master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Rename policy to policy_name, add AutoscalingSnapshotError(METRICS_UNAVAILABLE) and error logging

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Add DeploymentStatusTrigger enums

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
standardize metrics_health format to ms/s

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Refresh autoscaling deployments cache on app/deployment state changes

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun committed:
Fix incorrect display of custom autoscaling policy name

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun committed:
Clean up autoscaling_state.py by removing unnecessary casts and fixing formatting

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>
(cherry picked from commit 3e00556d2eb75a71c0d46ca243f95f7708a9a959)

nadongjun committed:
Clean up common.py by extracting autoscaling enums and improving type clarity

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>
(cherry picked from commit 3b36f3b1459af323f845394be163a21e6c1405ab)

nadongjun committed:
Add to_log_dict to AutoscalingDecisionSummary

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>
(cherry picked from commit c8294ac0a27ef80e71786e4e5243b321089cda40)

nadongjun committed:
Remove summarize_decisions and add to_log_dict to DecisionRecord for simpler logging

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>
(cherry picked from commit d899a994e7b69c6eb2a2d1ebab720ad4faf48140)

nadongjun commented on python/ray/serve/_private/common.py:
```diff
@@ -755,6 +762,136 @@ class CreatePlacementGroupRequest:
+        return f"{val:.2f}s"
+
+    @staticmethod
+    def summarize_decisions(
```
Replaced it with slicing as you suggested and removed the method.

nadongjun committed:
Remove AutoscalingDecisionSummary

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented on python/ray/serve/_private/common.py:
```diff
@@ -755,6 +762,136 @@ class CreatePlacementGroupRequest:
+                    "from": d.prev_num_replicas,
+                    "to": d.curr_num_replicas,
+                    "reason": d.reason,
+                }
```
`AutoscalingDecisionSummary` and `DecisionRecord` seemed to have overlapping roles, so I removed `AutoscalingDecisionSummary` and moved `to_log_dict()` to `DecisionRecord` instead.

Let me know if you think there’s value in keeping them separate.

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun committed:
apply review comments

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
improve autoscaling snapshot test

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

akyang-anyscale commented on python/ray/serve/_private/common.py:
```diff
@@ -755,6 +758,110 @@ class CreatePlacementGroupRequest:
+@dataclass
+class DecisionRecord:
+    timestamp_s: str
+    prev_num_replicas: int
+    curr_num_replicas: int
```
nit
```suggestion
    curr_num_replicas: int
    target_num_replicas: int
```

akyang-anyscale reviewed (commented):
a few minor comments, but overall lgtm!

nadongjun committed:
rename field name

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
change test timeout

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented on python/ray/serve/tests/test_controller.py:
```diff
@@ -276,6 +278,141 @@ class App:
+            return False
+
+    # Wait up to ~60s for the snapshot to appear.
+    wait_for_condition(_scan_for_snapshot, timeout=60)
```
It doesn’t need 60s. Other Serve tests use 15s, so I aligned this test accordingly. Thanks for pointing it out!

nadongjun committed:
Update python/ray/serve/_private/application_state.py

Co-authored-by: Abrar Sheikh <abrar2002as@gmail.com>
Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Update python/ray/serve/_private/application_state.py

Co-authored-by: Abrar Sheikh <abrar2002as@gmail.com>
Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
update autoscaling timestamp format and cache metrics timestamp

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
access Pydantic config fields directly

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

abrarsheikh commented on python/ray/serve/_private/autoscaling_state.py:
```diff
@@ -300,6 +332,81 @@ def get_total_num_requests(self) -> float:
+            min_replicas=min_replicas,
+            max_replicas=max_replicas,
+            scaling_status=scaling_status,
+            policy_name=getattr(ctx.config.policy, "name", None),
```
can we do this?

```suggestion
            policy_name=ctx.config.policy.name,
```

nadongjun committed:
Update python/ray/serve/_private/autoscaling_state.py

Co-authored-by: Abrar Sheikh <abrar2002as@gmail.com>
Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Add is_scaling_equivalent for snapshot deduplication and test no logs when autoscaling_config is missing

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented on python/ray/serve/_private/common.py:
```diff
@@ -755,6 +758,110 @@ class CreatePlacementGroupRequest:
+            and self.total_requests == other.total_requests
+        )
+
+    def to_log_dict(self) -> Dict[str, object]:
```
Pydantic dataclasses have no built-in to_dict. dataclasses.asdict() can convert all fields, but selecting or renaming fields still requires a custom method such as `to_log_dict`.

abrarsheikh replied:
i dont think this custom dict transformation is not desirable, because you plan to use this dict to write a structured log. Then, from a different tool/process, deserialize the structured log to construct this same Pydantic object again. By creating a custom serialization method like `to_log_dict`, you lose the ability to reconstruct the dataclass automatically. consider the following instead


```
# --- logging ---
snap = DeploymentSnapshot(
    timestamp=datetime.utcnow(),
    app="myapp",
    deployment="router",
    current_replicas=2,
    target_replicas=3,
    min_replicas=None,
    max_replicas=10,
    scaling_status="scaling_up",
    policy_name="pid_v1",
    look_back_period_s=30.0,
    queued_requests=4.0,
    total_requests=123.0,
    metrics_health="healthy",
    errors=[],
    decisions=[],
)

# Structured log as JSON (stable schema, excludes None):
log_json = snap.model_dump_json(by_alias=True, exclude_none=True)

# --- reconstruction elsewhere ---
snap2 = DeploymentSnapshot.model_validate_json(log_json)
assert snap2 == snap
```


To a large degree this also protects from schema changes to the dataclass in the future, which you dont get if you write your own serialization.

nadongjun replied:
agree, I dropped to_log_dict and now use model_dump_json / model_validate_json for structured logging and reconstruction as you suggested

nadongjun commented on python/ray/serve/_private/common.py:
```diff
@@ -755,6 +758,110 @@ class CreatePlacementGroupRequest:
+    errors: List[str]
+    decisions: List[DecisionRecord]
+
+    def __eq__(self, other: object) -> bool:
```
I removed the `__eq__` implementation on DeploymentSnapshot and introduced a new method `is_scaling_equivalent`. This makes the comparison logic explicit and only checks scaling-related fields (`current_replicas` and `scaling_status`) when deciding whether to emit a new autoscaling snapshot log.

nadongjun committed:
Merge remote-tracking branch 'origin/master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

abrarsheikh commented on python/ray/serve/tests/test_controller.py:
```diff
@@ -276,6 +278,201 @@ class App:
     assert isinstance(deployment_config, DeploymentConfig)
 
 
+def test_autoscaling_snapshot_log_emitted_and_well_formed(serve_instance):
```
if the intention is to read this log back from controller logs to serve the cli, I think that is a suboptimal design. I would instead write the autoscaling snapshots to a different log file

nadongjun replied:
agree, I updated the design so that autoscaling snapshots are now written to a dedicated rotating log file (autoscaling_snapshot_<pid>.log) instead of being mixed into the controller logs. This keeps controller logs clean and makes it easier for external tools (like the CLI) to consume snapshot data directly

<img width="1082" height="150" alt="image" src="https://github.com/user-attachments/assets/ae27fe53-4d7d-450a-ac01-1ed55a9757f7" />

nadongjun committed:
refactor DeploymentSnapshot models to use Pydantic

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
add dedicated rotating file logger for autoscaling snapshots

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented on python/ray/serve/_private/autoscaling_state.py:
```diff
@@ -589,6 +619,81 @@ def get_total_num_requests(self) -> float:
+        scaling_status = DeploymentSnapshot.format_scaling_status(scaling_status_raw)
+
+        look_back_period_s = self._config.look_back_period_s
+        metrics_health = DeploymentSnapshot.format_metrics_health_text(
```
`metrics_health` shows how long it’s been since the last metric (time.time() - self._latest_metrics_timestamp), not a string status. It returns unknown or the metric age, which we’ve found more useful for debugging than simple strings like "ok". Do you think the name should be clearer?

abrarsheikh commented on python/ray/serve/_private/common.py:
```diff
@@ -755,6 +759,77 @@ class CreatePlacementGroupRequest:
+    policy_name: str
+    look_back_period_s: Optional[float]
+    queued_requests: Optional[float]
+    total_requests: float
```
this is to standardize the naming across the repo: queued_requests, running_requests and ongoing_requests. ongoing_requests = queued_requests + running_requests

```suggestion
    ongoing_requests: float
```

nadongjun committed:
rename total_requests to ongoing_requests

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
update testcode

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
add structured JSON logging for autoscaling snapshots

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
add snapshot type

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
use DeploymentSnapshot type instead of object

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
use timestamp_str, app, and deployment for scaling equivalence

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
refactor autoscaling snapshot test with simplified

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented on python/ray/serve/_private/logging_utils.py:
```diff
@@ -412,6 +412,42 @@ def configure_component_logger(
+    handler = logging.handlers.RotatingFileHandler(
+        file_path, maxBytes=max_bytes, backupCount=backup_count
+    )
+    handler.setFormatter(ServeFormatter("autoscaling_snapshot", component_id))
```
Got it, I updated it so we can parse the logs directly with JSON, like this.
```py
import json

with open("autoscaling_snapshot_78110.log", "r", encoding="utf-8") as f:
    for line in f:
        try:
            record = json.loads(line)
        except json.JSONDecodeError:
            continue

        snapshot = record.get("snapshot", {})
        print(json.dumps(snapshot, ensure_ascii=False, indent=2))
```

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun committed:
update testcode

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
update testcode snapshot key

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
update return type

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge remote-tracking branch 'origin/master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Update sum handle queue in ctx & snapshot policy class path

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Remove sleep

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Remove exception handling

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Use get_serve_logs_dir

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge remote-tracking branch 'origin/master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

akyang-anyscale commented on python/ray/serve/_private/autoscaling_state.py:
```diff
@@ -557,6 +589,77 @@ def get_total_num_requests(self) -> float:
+        else:
+            scaling_status_raw = AutoscalingStatus.STABLE
+
+        scaling_status = DeploymentSnapshot.format_scaling_status(scaling_status_raw)
```
this method `format_scaling_status` seems like it should be a method of `AutoscalingStatus`

akyang-anyscale commented on python/ray/serve/_private/common.py:
```diff
@@ -755,6 +759,78 @@ class CreatePlacementGroupRequest:
+    ongoing_requests: float
+    metrics_health: str
+    errors: List[str]
+    decisions: List[DecisionRecord]
```
is there a fundamental need for `DecisionRecord` given `DeploymentSnapshot` seems to be the same + more details?

akyang-anyscale commented on python/ray/serve/_private/controller.py:
```diff
@@ -400,6 +414,47 @@ def _update_proxy_nodes(self):
+            dep_name,
+            details,
+            autoscaling_config,
+        ) in self._list_deployments_for_autoscaling():
```
```suggestion
        ) in self._autoscaling_enabled_deployments_cache:
```

nadongjun committed:
Add return type hint and minor condition check

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Refactor move scaling status formatting to AutoscalingStatus

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun committed:
edit

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented:
@akyang-anyscale `DeploymentSnapshot` is generated almost every second, while `DecisionRecord` summarizes each scaling event. This makes debugging much easier, without it, the reason behind changes gets buried in frequent snapshots.

Two questions:
1. Currently, `is_scaling_equivalent()` only compares timestamp_str/app/deployment, so dedup rarely triggers.
Would it make more sense to compare scaling-related fields (current/target/min/max replicas, scaling_status, etc.) instead?

2. Autoscaling snapshot logs are currently written only to log files and never shown on Serve runtime stdout.
Would it be better to show key events (scale up/down, errors, etc.) on stdout during runtime, while keeping full snapshots in log files?

akyang-anyscale commented on python/ray/serve/_private/autoscaling_state.py:
```diff
@@ -232,7 +245,22 @@ def get_decision_num_replicas(
+                timestamp_str=time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
+                current_num_replicas=autoscaling_context.current_num_replicas,
+                target_num_replicas=decision_num_replicas,
+                reason=f"current={autoscaling_context.current_num_replicas}, target={decision_num_replicas}",
```
what's the purpose of the reason field?

akyang-anyscale commented:
> Currently, is_scaling_equivalent() only compares timestamp_str/app/deployment, so dedup rarely triggers.
Would it make more sense to compare scaling-related fields (current/target/min/max replicas, scaling_status, etc.) instead?

Yes we should only log snapshots that had a meaningful change (i.e. the target changed)

> Autoscaling snapshot logs are currently written only to log files and never shown on Serve runtime stdout.
Would it be better to show key events (scale up/down, errors, etc.) on stdout during runtime, while keeping full snapshots in log files?

That makes sense to log key events to stdout

nadongjun committed:
Refactor autoscaling snapshot dedup logic and remove reason field

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun committed:
Merge remote-tracking branch 'origin/master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun commented:
Hi @akyang-anyscale 
I’m curious about the plan for this PR.
If any further changes are needed, I’m happy to make them.

nadongjun committed:
Merge remote-tracking branch 'origin/master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

akyang-anyscale reviewed (approved):
lgtm!

abrarsheikh commented on python/ray/serve/_private/application_state.py:
```diff
@@ -1393,8 +1397,12 @@ def get_deployment_topology(self, app_name: str) -> Optional[DeploymentTopology]
+    def update(self) -> bool:
+        """
+        Update each application state.
+        Returns:
```
```suggestion
        Update each application state.

        Returns:
```

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -476,12 +528,18 @@ async def run_control_loop_step(
 
+        try:
+            # Emit one autoscaling snapshot per deployment per loop using existing state.
+            self._emit_deployment_autoscaling_snapshots()
```
does this write to file on every controller iteration ?

abrarsheikh commented on python/ray/serve/_private/controller.py:
```diff
@@ -395,6 +409,47 @@ def _update_proxy_nodes(self):
+            payload = deployment_snapshot.model_dump(exclude_none=True)
+            self._autoscaling_logger.info(
+                "", extra={"type": "deployment", "snapshot": payload}
+            )
```
`payload` is already json because of `model_dump`. And type should be part of `deployment_snapshot` object in my opinion.

the extra argument to logger.info is used in a non traditional way here IMO

nadongjun committed:
Merge remote-tracking branch 'origin/master' into serve-obsv-deployment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Update python/ray/serve/_private/application_state.py

Co-authored-by: Abrar Sheikh <abrar2002as@gmail.com>
Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Edit autoscaling snapshot logging and add deduplication test

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Remove decisions history from DeploymentSnapshot

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Fix Pydantic compatibility

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Fix memory leak

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

abrarsheikh commented on python/ray/serve/_private/logging_utils.py:
```diff
@@ -412,6 +412,42 @@ def configure_component_logger(
 
 
+# Configure a dedicated rotating file logger for autoscaling snapshots.
+def configure_snapshot_logger(
```
```suggestion
def configure_autoscaling_snapshot_logger(
```

nadongjun committed:
Remove unused DecisionRecord and clarify autoscaling snapshot logger init

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

abrarsheikh commented on python/ray/serve/_private/autoscaling_state.py:
```diff
@@ -243,7 +250,13 @@ def get_decision_num_replicas(
+        self._cached_deployment_snapshot = self._create_deployment_snapshot(
+            ctx=autoscaling_context,
+            target_replicas=decision_num_replicas,
+        )
```
think this is okay for now because we will handle it in the future. worth add a TODO comment

nadongjun committed:
Batch autoscaling snapshot logs per control loop and update tests

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Fix autoscaling metrics timestamp regression

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Add comment

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Update test_autoscaling_snapshot_batched_single_write_per_loop

Signed-off-by: Dongjun Na <kmu5544616@gmail.com>

nadongjun committed:
Merge branch 'master' into serve-obsv-deployment

nadongjun commented:
> tests are failing

Fixed the failing tests!
